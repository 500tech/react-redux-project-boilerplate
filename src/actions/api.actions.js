// @flow
import { set } from 'lodash/fp';
import type { ApiAction } from 'types/redux.types';

export const apiAction = (action: ApiAction): ApiAction =>
  set('meta.api', true, action);
