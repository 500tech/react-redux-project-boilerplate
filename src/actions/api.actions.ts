import { BaseAction } from 'types/base-redux.types';

export type HttpMethod =
  | 'get'
  | 'post'
  | 'put'
  | 'delete'
  | 'path'
  | 'head'
  | 'options';

type OnSuccess<ResponseBody> = (data: ResponseBody) => BaseAction;
type OnError = (error: any) => void;

export interface ApiAction<ResponseBody> extends BaseAction {
  meta: {
    api: true;
  };
  payload: {
    networkLabel: string;
    method: HttpMethod;
    path: string;
    data?: any;
    baseUrl?: string;
    onSuccess: OnSuccess<ResponseBody> | OnSuccess<ResponseBody>[];
    onError?: OnError | OnError[];
  };
}
