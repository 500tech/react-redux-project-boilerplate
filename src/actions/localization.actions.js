// @flow
// TODO: remove this file if no localization is needed
import type { BaseAction } from 'types/redux.types';

export const SET_LOCALE: string = '[localization] Set Locale';

export type SetLocaleAction = BaseAction<{|
  +locale: string
|}>;
export type LocalizationAction = SetLocaleAction;

export const setLocale = (locale: string): SetLocaleAction => ({
  type: SET_LOCALE,
  payload: {
    locale
  }
});
