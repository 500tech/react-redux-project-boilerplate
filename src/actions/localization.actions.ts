// TODO: remove this file if no localization is needed
import { BaseAction } from 'types/base-redux.types';

export const SET_LOCALE: string = '[localization] Set Locale';

export interface SetLocaleActionPayload {
  locale: string;
}

export type SetLocaleAction = BaseAction<SetLocaleActionPayload>;
export type LocalizationAction = SetLocaleAction;

export const setLocale = (locale: string): SetLocaleAction => ({
  type: SET_LOCALE,
  payload: {
    locale
  }
});
