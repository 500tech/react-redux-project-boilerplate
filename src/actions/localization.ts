// TODO: remove this file if no localization is needed
import { BaseAction } from 'types/base-redux';
import { LocaleTypes } from 'constants/locales';

export const SET_LOCALE: string = '[localization] Set Locale';

export interface SetLocaleAction extends BaseAction {
  payload: {
    locale: string;
  };
}

export const setLocale = (locale: LocaleTypes) => ({
  type: SET_LOCALE,
  payload: {
    locale
  }
});
