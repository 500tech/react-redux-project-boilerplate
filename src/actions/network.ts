import { BaseAction } from 'types/base-redux';

export const START_NETWORK: string = '[network] Start';
export const END_NETWORK: string = '[network] End';

export interface NetworkAction extends BaseAction {
  payload: {
    networkLabel: string;
  };
}

export const startNetwork = (networkLabel: string) => ({
  type: START_NETWORK,
  payload: {
    networkLabel
  }
});

export const endNetwork = (networkLabel: string): NetworkAction => ({
  type: END_NETWORK,
  payload: {
    networkLabel
  }
});
