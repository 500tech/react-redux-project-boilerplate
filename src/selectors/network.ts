import { createSelector } from 'reselect';
import { some, castArray } from 'lodash/fp';

import { State } from 'types/redux';

// TODO: make into selector factory
export const makeIsLoadingSelector = (networkLabel: string | string[]) =>
  createSelector(
    (state: State) => state.network,
    (network) => {
      const labels = castArray(networkLabel);

      return some((currentLabel) => network[currentLabel] > 0, labels);
    }
  );
