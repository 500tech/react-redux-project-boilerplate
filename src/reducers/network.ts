import { set } from 'lodash/fp';
import { handleActions } from 'redux-actions';

import * as AT from 'actions/network';

export type NetworkState = Record<string, number>;

const initialState: NetworkState = {};

const network = handleActions<NetworkState>(
  {
    [AT.START_NETWORK](state, action: ReturnType<typeof AT.startNetwork>) {
      const { networkLabel } = action.payload;

      return set(
        networkLabel,
        state[networkLabel] ? state[networkLabel] + 1 : 1,
        state
      );
    },
    [AT.END_NETWORK](state, action: ReturnType<typeof AT.endNetwork>) {
      const { networkLabel } = action.payload;

      return set(
        networkLabel,
        state[networkLabel] ? state[networkLabel] - 1 : 0,
        state
      );
    }
  },
  initialState
);

export default network;
