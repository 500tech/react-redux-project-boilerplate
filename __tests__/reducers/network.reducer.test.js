import networkReducer from 'reducers/network.reducer';
import { endNetwork, startNetwork } from 'actions/network.actions';

describe('Reducer: Network', () => {
  test('startNetwork: should save pending request', () => {
    const action = startNetwork('posts');
    const result = networkReducer(undefined, action);

    expect(result.posts).toEqual(1);
    expect(result).toMatchSnapshot();
  });

  test('endNetwork: should remove finished request', () => {
    const action = endNetwork('posts');
    const initialState = {
      posts: 1
    };
    const result = networkReducer(initialState, action);

    expect(result.posts).toEqual(0);
    expect(result).toMatchSnapshot();
  });

  test('startNetwork: should increment pending request with same label', () => {
    const action = startNetwork('posts');
    const initialState = {
      posts: 1
    };

    const secondResult = networkReducer(initialState, action);
    expect(secondResult.posts).toEqual(2);
  });

  test('endNetwork: should decrement a request with the same label', () => {
    const action = endNetwork('posts');
    const initialState = {
      posts: 2
    };

    const secondResult = networkReducer(initialState, action);
    expect(secondResult.posts).toEqual(1);
  });

  test('endNetwork: should not decrement a label below 0', () => {
    const action = endNetwork('posts');
    const initialState = {
      posts: 0
    };

    const secondResult = networkReducer(initialState, action);
    expect(secondResult.posts).toEqual(0);
  });
});
