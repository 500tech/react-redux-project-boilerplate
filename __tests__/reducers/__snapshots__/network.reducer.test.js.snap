// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer: Network endNetwork: should remove finished request 1`] = `
Object {
  "posts": 0,
}
`;

exports[`Reducer: Network startNetwork: should save pending request 1`] = `
Object {
  "posts": 1,
}
`;
